src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -162,38 +204,119 @@ pub struct ObligationForest<O: ForestObligation> {
     ///
     /// [details]: https://github.com/rust-lang/rust/pull/53255#issuecomment-421184780
     error_cache: FxHashMap<ObligationTreeId, FxHashSet<O::CacheKey>>,
+
+    /// Stores which nodes would be unblocked once `O::Variable` is unified
+    stalled_on: FxHashMap<O::Variable, Vec<NodeIndex>>,
+    /// Stores the node indices that are unblocked and should be processed at the next opportunity
+    unblocked: BinaryHeap<Unblocked>,
+    /// Stores nodes which should be processed on the next iteration since the variables they are
+    /// actually blocked on are unknown
+    stalled_on_unknown: Vec<NodeIndex>,
+    /// The offset that this `ObligationForest` has registered. Should be de-registered before
+    /// dropping this forest.
+    offset: Option<O::WatcherOffset>,
+    /// Reusable vector for storing unblocked nodes whose watch should be removed
+    temp_unblocked_nodes: Vec<O::Variable>,
+}
+
+/// Helper struct for use with `BinaryHeap` to process nodes in the order that they were added to
+/// the forest
+struct Unblocked {
+    index: NodeIndex,
+    order: u32,
+}
+
+impl PartialEq for Unblocked {
+    fn eq(&self, other: &Self) -> bool {
+        self.order == other.order
+    }
+}
+impl Eq for Unblocked {}
+impl PartialOrd for Unblocked {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        other.order.partial_cmp(&self.order)
+    }
+}
+impl Ord for Unblocked {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.order.cmp(&self.order)
+    }
 }
 
 #[derive(Debug)]
-struct Node<O> {
+struct Node<O: ForestObligation> {
     obligation: O,
     state: Cell<NodeState>,
 
+    /// A predicate (and its key) can changed during processing. If it does we need to register the
```

Comment:
```suggestion
    /// A predicate (and its key) can change during processing. If it does we need to register the
```

Reply:
Might be nice to clarify *how* or *when* a predicate can change, but this is very helpful.

Reply:
I *think* this is because variables gets instantiated, however I am not 100% on that. The comment mostly echoes another, existing comment about predicates changing when `process_obligation` is called (which also does not explain how a predicate changes)